htdocs/survey_react.php

```diff
@@ -0,0 +1,336 @@
+
+        try {
+            $this->Instrument = \NDB_BVL_Instrument::factory(
+                $this->SurveyInfo['Test_name'],
+                $this->SurveyInfo['CommentID'],
+                null,
+                true
+            );
+        } catch(\Exception $e) {
+            throw new Exception("Instrument not found", 405);
+        }
+
+        $fp   = fopen("php://input", "r");
+        $data = '';
+        while (!feof($fp)) {
+            $data .= fread($fp, 1024);
+        }
+        fclose($fp);
+
+        $data            = json_decode($data);
+        $instrument_name = $this->Instrument->testName;
+
+        if(count($data) !== 1) {
+            // The survey module will only PATCH one variable at a time.
+            header("HTTP/1.0 400 Bad Request");
+        }
+
+        if ($this->Instrument->validate($data)) {
+            try {
+                $this->Instrument->_save($data);
+                
+            } catch (Exception $e) {
+                header("HTTP/1.0 400 Bad Request");
+            }
+        } else {
+            $this->Header("HTTP/1.1 403 Forbidden");
+            if (!$this->Instrument->determineDataEntryAllowed()) {
+                $msg = "Can not update instruments that"
+                       . " are flagged as complete";
+                // $this->JSON = array('error' => $msg);
+            } else {
+                // $this->JSON = array("error" => "Could not update.");
+            }
+        }
+    }
+
+    /**
+     * Handle a PATCH request. This will update a single field
+     *
+     * @return none
+     */
+    function handlePUT() {
+
+        $fp   = fopen("php://input", "r");
+        $data = '';
+        while (!feof($fp)) {
+            $data .= fread($fp, 1024);
+        }
+        fclose($fp);
+
+        $data = json_decode($data, true);
+        $subtest = null;
+
+        if($data['page'] > 0) {
+            $subtest = $this->SurveyInfo['Test_name'] . "_page" . $data['page'];
+        }
+
+        try {
+            $this->Instrument = \NDB_BVL_Instrument::factory(
+                $this->SurveyInfo['Test_name'],
+                $this->SurveyInfo['CommentID'],
+                $subtest,
+                true
+            );
+        } catch(\Exception $e) {
+            throw new Exception("Instrument not found", 405);
+        }
+
+        $this->Instrument->form->directEntry = true;
+        $this->Instrument->form->directValues = $data['data'];
+
+        if($this->Instrument->form->validate()) {
+            if($data['FinalPage']) {
+                echo $this->Instrument->getReactReview();
+            } else {
+                echo $this->Instrument->toDirectJSON();
+            }
+        } else {
+            header("HTTP/1.0 400 Bad Request");
+            echo json_encode($this->Instrument->form->errors);
+        }
+
+    }  
+
+    /**
+     * Handle a PATCH request. This will update a single field
+     *
+     * @return none
+     */
+    function handlePOST() {
```

Comment:
Used to submit an instrument. Checks all fields are valid and marks complete if so